app/components/DocumentContext.tsx

```diff
@@ -0,0 +1,76 @@
+import { action, computed, observable } from "mobx";
+import React, { PropsWithChildren } from "react";
+import { Heading } from "@shared/utils/ProsemirrorHelper";
+import Document from "~/models/Document";
+import { Editor } from "~/editor";
+
+class DocumentContext {
```

Comment:
Having the document and editor instance here makes it easy to central the editor event updates.
One potential problem is, the context is provided high in the component tree (in AuthenticatedLayout). We can introduce another context for the state, but we'll end up duplicating the editor references across two contexts.